react/lesson1.md

```diff
@@ -36,175 +91,431 @@ cd my-app/
+State can be accessed anywhere in the components using `this.state.myName` and can be set using `this.setState({myName: "hello"})`. 
 
-## Thinking in React
-In development you are often given specification of a piece work with example design and data:
+`setState` also tells React that our data has changed, so it can go ahead and re-render our view. Hence, it's very important that you never change state directly, and you only modify using `setState`:
 
-![Inventory data table](https://facebook.github.io/react/img/blog/thinking-in-react-mock.png)
+```javascript
+//WRONG
+this.state.myName = "hello";
 
-```json
-[
-  {"category": "Sporting Goods", "price": "$49.99", "stocked": true, "name": "Football"},
-  {"category": "Sporting Goods", "price": "$9.99", "stocked": true, "name": "Baseball"},
-  {"category": "Sporting Goods", "price": "$29.99", "stocked": false, "name": "Basketball"},
-  {"category": "Electronics", "price": "$99.99", "stocked": true, "name": "iPod Touch"},
-  {"category": "Electronics", "price": "$399.99", "stocked": false, "name": "iPhone 5"},
-  {"category": "Electronics", "price": "$199.99", "stocked": true, "name": "Nexus 7"}
-];
+//CORRECT
+this.setState({
+  myName: "hello"
+});
 ```
-Steps to conceptualise the solution with `React`:
-- 1: [Break The UI Into A Component Hierarchy](https://facebook.github.io/react/img/blog/thinking-in-react-components.png)
-- 2: Build A Static Version
-- 3: Identify the State and where it should live
 
-[Exercise: Code Review](http://codepen.io/lacker/pen/vXpAgj)
+Remember the change handler we added for our difficulty setting? And the array that builds our buttons array? Time to connect those two using state!
 
-# JSX: Map Arrays to Components
+Start by defining the initial state in the constructor:
 
-When working with APIs we often have collections of things in `Arrays` which we want to map to components. This means we can loop through that data and display the customisable `dumb` component for each one.
+```javascript
+constructor() {
+  super();
+  this.state = {
+    difficulty: 10
+  };
+}
+```
 
-```jsx
+Notice the call to `super()`. We always need to call that before referencing the `this` keyword in the constructor. In this case it calls the `constructor()` of the parent class `React.Component`. `super` is not a React function, it's an ES6 keyword used in class inherritence: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/super
 
-renderOrganisations() {
-    const orgData = [{ name: 'Organisation 1', borough: 'Camden'}, {name: 'Organisation 2'}];
-    return orgData.map(function renderData(organisation) {
-        return (
-            <Organisation
-                name= {organisation.name}
-                borough={organisation.borough ? organisation.borough : 'None'}
-                website= {organisation.website}
-            />
-        )
+```javascript
+  difficultyChange = (event) => {
```

Comment:
Sidenote: this is not currently in ES6 - `create-react-app` will give us [babel-plugin-transform-class-properties](https://www.npmjs.com/package/babel-plugin-transform-class-properties) out of the box.

Not sure if it's worth mentioning - downside is that the alternatives are more complicated to understand (either `.bind(this` in `render` or in the `constructor`)